graphql_compiler/macros/macro_expansion.py

```diff
@@ -47,75 +67,105 @@ def _merge_selection_sets(selection_set_a, selection_set_b):
+    # having an empty selection in a vertex field. We remove pro-forma fields if they are
+    # no longer necessary.
+    if len(merged_selection_dict) > 1:  # Otherwise we need a pro-forma field
         merged_selection_dict = {
             name: ast
             for name, ast in six.iteritems(merged_selection_dict)
             if ast.selection_set is not None or len(ast.directives) > 0
+            # If there's selections or directives under the field, it is not pro-forma.
         }
 
     # Get a deterministic ordering of the merged selections
-    selection_name_order = [
+    selection_name_order = list(chain((
         ast.name.value
         for ast in selection_set_a.selections
         if ast.name.value not in selection_dict_b
-    ] + [
+    ), (
         ast.name.value
         for ast in selection_set_b.selections
-    ]
+    )))
 
     # Make sure that all property fields come before all vertex fields.
-    return SelectionSet([
+    return SelectionSet(list(chain((
         merged_selection_dict[name]
         for name in selection_name_order
         if name in merged_selection_dict and merged_selection_dict[name].selection_set is None
-    ] + [
+    ), (
         merged_selection_dict[name]
         for name in selection_name_order
         if name in merged_selection_dict and merged_selection_dict[name].selection_set is not None
-    ])
+    ))))
 
 
 def _merge_selection_into_target(target_ast, selection_ast, subclass_sets=None):
-    """Add the selections, directives, and coercions from the selection_ast to the target_ast."""
+    """Add the selections, directives, and coercions from the selection_ast to the target_ast.
+
+    Mutate the target_ast, merging into it everything from the selection_ast. If the target
+    is at a type coercion and the selection_ast starts with a type coercion, combine them
+    into one coercion that preserves the semantics of nested coercions, which are disallowed.
+
+    For details on how fields and directives are merged, see _merge_selection_sets.
+
+    Args:
+        target_ast: AST at the @macro_edge_target directive
+        selection_ast: AST to merge inside the target. Required to have a nonempty selection set.
+        subclass_sets: optional dict mapping class names to the set of its subclass names
+    """
+    if selection_ast.selection_set is None or not selection_ast.selection_set.selections:
+        raise AssertionError(u'Precondition violated. selection_ast is expected to be nonempty {}'
+                             .format(selection_ast))
+
+    # See if there's a type coercion in the selection_ast
+    coercion = None
+    for selection in selection_ast.selection_set.selections:
+        if isinstance(selection, InlineFragment):
+            if len(selection_ast.selection_set.selections) != 1:
+                raise GraphQLCompilationError(u'Found selections outside type coercion {}'
```

Comment:
This is a user-facing error, so let's give the user some actionable advice, and let's avoid printing a giant AST selection set in the error message.